perma_web/api2/serializers.py

```diff
@@ -0,0 +1,242 @@
+    registrar = serializers.StringRelatedField()
+    shared_folder = FolderSerializer()
+
+    class Meta:
+        model = Organization
+        fields = ('id', 'name', 'registrar', 'default_to_private', 'shared_folder')
+
+
+### CAPTUREJOB ###
+
+class CaptureJobSerializer(BaseSerializer):
+    guid = serializers.PrimaryKeyRelatedField(source='link', read_only=True)
+    class Meta:
+        model = CaptureJob
+        fields = ('guid', 'status', 'attempt', 'step_count', 'step_description', 'capture_start_time', 'capture_end_time', 'queue_position')
+
+
+### CAPTURE ###
+
+class CaptureSerializer(BaseSerializer):
+    class Meta:
+        model = Capture
+        fields = ('role', 'status', 'url', 'record_type', 'content_type', 'user_upload', 'playback_url')
+
+
+### LINK ###
+
+class LinkSerializer(BaseSerializer):
+    url = serializers.CharField(source='submitted_url', max_length=2100, required=False, allow_blank=True)
+    title = serializers.CharField(source='submitted_title', max_length=2100, required=False)
+    description = serializers.CharField(source='submitted_description', allow_blank=True, allow_null=True, max_length=300, required=False)
+
+    captures = CaptureSerializer(many=True, read_only=True)
+    queue_time = serializers.SerializerMethodField()
+    capture_time = serializers.SerializerMethodField()
+
+    class Meta:
+        model = Link
+        fields = ('guid', 'creation_timestamp', 'url', 'title', 'description', 'warc_size', 'captures', 'queue_time', 'capture_time')
+
+    def get_queue_time(self, link):
+        try:
+            delta = link.capture_job.capture_start_time - link.creation_timestamp
+            return delta.seconds
+        except:
+            return None
+
+    def get_capture_time(self, link):
+        try:
+            delta = link.capture_job.capture_end_time - link.capture_job.capture_start_time
+            return delta.seconds
+        except:
+            return None
+
+
+class AuthenticatedLinkSerializer(LinkSerializer):
+    """
+        Subclass of LinkSerializer with extra fields for users who own the link.
+    """
+    created_by = LinkUserSerializer(read_only=True)
+    organization = OrganizationSerializer(read_only=True)
+
+    class Meta(LinkSerializer.Meta):
+        fields = LinkSerializer.Meta.fields + ('notes', 'created_by', 'is_private', 'private_reason', 'archive_timestamp', 'organization')
+        allowed_update_fields = ['submitted_title', 'submitted_description', 'notes', 'is_private', 'private_reason']
+
+    def validate_url(self, url):
+        # Clean up the user submitted url
+        url = url.strip()
+        if url and url[:4] != 'http':
+            url = 'http://' + url
+        return url
+
+    def validate(self, data):
+        user = self.context['request'].user
+        errors = {}
+
+        # handle private_reason:
+        if not user.is_staff:
+            # only staff can manually set private_reason
+            data.pop('private_reason', None)
```

Comment:
Triviality: does it make sense for this popping off of 'private_reason') to take place inside the "if self.instance" branch, to make it clear that staff can only affect the private_reason when updating an existing link? I think probably yes. If I'm reading this correctly, any private_reason provided by a staff member for a not-yet-saved link would be overwritten in the "else" below.